pkg/acs/calacs/acsccd/doccd.c

```diff
@@ -143,278 +171,582 @@ int DoCCD (ACSInfo *acs_info) {
 
-    if (!OmitStep(acs[0].biascorr))
-        if (biasHistory(&acs[0], x[0].globalhdr))
+    if (!OmitStep(acs[primaryIdx].biascorr)) {
+        if (biasHistory(&acs[primaryIdx], x[primaryIdx].globalhdr)) {
+            freeOnExit (&ptrReg);
             return (status);
+       }
+    }
     /************************************************************************/
 
     /************************************************************************/
     /* convert data to electrons */
+    {unsigned int i;
     for (i = 0; i < acs_info->nimsets; i++) {
         if (to_electrons(&acs[i], &x[i])) {
-            freeSingleGroup(&x[i]);
+            freeOnExit (&ptrReg);
             return (status);
         }
 
         if (PutKeyStr(&x[i].sci.hdr, "BUNIT", "ELECTRONS", "")) {
-            freeSingleGroup(&x[i]);
+            freeOnExit (&ptrReg);
             return (status);
         }
         if (PutKeyStr(&x[i].err.hdr, "BUNIT", "ELECTRONS", "")) {
-            freeSingleGroup(&x[i]);
+            freeOnExit (&ptrReg);
             return (status);
         }
-    }
+    }}
     /************************************************************************/
 
     /************************************************************************/
     /* Subtract bias level determined from overscan. */
-    BlevMsg(&acs[0], 1);
+    BlevMsg(&acs[primaryIdx], nextver);
 
-    for (i = 0; i < acs_info->nimsets; i++) {
-        blevcorr[i] = PERFORM;
+    int * blevcorr = NULL;
+    blevcorr = (int *) malloc(acs_info->nimsets * sizeof(int));
+    if (!blevcorr) {
+       freeOnExit (&ptrReg);
+       return (OUT_OF_MEMORY);
     }
+    addPtr (&ptrReg, blevcorr, free);
 
-    /* default blev case, do original bias level subtraction */
-    if (acs_info->blevcorr == PERFORM &&
-        ((acs_info->expstart < SM4MJD && acs_info->detector == WFC_CCD_DETECTOR) ||
-         (acs_info->detector == WFC_CCD_DETECTOR && acs_info->subarray == YES) ||
-         (acs_info->detector != WFC_CCD_DETECTOR))) {
-        for (i = 0; i < acs_info->nimsets; i++) {
-            /* This is set based on results of FindOver */
-            done = overscan[i];
-
-            if (doBlev(&acs[i], &x[i], acs[i].chip, &meanblev, &done,
-                       &driftcorr))
-                return (status);
-
-            if (done) {
-                trlmessage("Bias level from overscan has been subtracted;");
+    {unsigned int i;
+    for (i = 0; i < acs_info->nimsets; i++) {
+        blevcorr[i] = PERFORM;
+    }}
+
+    /* The logic here has been re-written (1) to accommodate the new subarrays
+       (Ref: ISR ACS 2017-03) which can be bias shift corrected ("*-2K" data only at 
+       this time), and (2) the head of the ACS team requested the logic be clarified.  
+       This code is verbose and potentially harder to maintain, but the logic is clearer
+       for the scientist reading the code.
+
+       NOTE: The variable "done" is overloaded.  When done = overscan[i], this means
```

Comment:
Can't we un-overload and instead of any ``if (done)`` use if (blevDone || overscan[i])``? Not 100% certain of ``||`` vs ``&&`` as I haven't read any of the below logic.